### PR TITLE
TextField border gap padding improvement

### LINKED ISSUE #82321: TextField with rounded corners - floating label shows extra whitespace
This is a similar issue as #33948.

I need my `TextField` (or `TextFormField`) to have rounded corners in order to match my company's design:

(picture taken from Vue.js application)
Screenshot 2021-05-12 at 08 29 33

I successfully achieved it by setting all the borders of the `TextField`'s `InputDecoration` to:
```dart
OutlineInputBorder(
  borderRadius: BorderRadius.circular(32.0),
)
```

However, the problem I face is that the floating label (which is especially needed for the desired design) is looking strange in my Flutter app:

Screenshot 2021-05-12 at 08 25 17

It has too much whitespace around it. When decreasing the border radius e.g. to `8.0`, the floating label displays fine. But I need it to be `32.0` in order to be completely round.

<details>
  <summary>Code to reproduce it</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  final border = OutlineInputBorder(
    borderRadius: BorderRadius.circular(32.0),
  );

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: Text("TextField test")),
        body: Container(
          alignment: Alignment.center,
          padding: const EdgeInsets.all(16.0),
          child: TextField(
            obscureText: true,
            decoration: InputDecoration(
              labelText: "Password",
              focusedBorder: border,
              enabledBorder: border,
              errorBorder: border,
              border: border,
              disabledBorder: border,
              focusedErrorBorder: border,
            ),
          ),
        ),
      ),
    );
  }
}
```

</details>

<details>
  <summary>flutter doctor</summary>

```
   Doctor summary (to see all details, run flutter doctor -v):
  [✓] Flutter (Channel stable, 2.0.6, on macOS 11.3.1 20E241 darwin-x64, locale en-GB)
  [✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
  [✓] Xcode - develop for iOS and macOS
  [✗] Chrome - develop for the web (Cannot find Chrome executable at /Applications/Google
      Chrome.app/Contents/MacOS/Google Chrome)
      ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.
  [✓] Android Studio (version 4.2)
  [✓] VS Code (version 1.56.0)
  [✓] Connected device (1 available)
  
  ! Doctor found issues in 1 category.
```
</details>

### PR DESCRIPTION
**Pull request by justinmc**
Previously, the calculations for the border gap weren't quite right in all cases, and would sometimes result in extra unwanted padding.  Here's an example where gapPadding is set to zero, meaning that the border should just touch the label, but it didn't:

```dart
          child: TextField(
            decoration: InputDecoration(
              labelText: 'Hello label',
              border: OutlineInputBorder(
                borderRadius: BorderRadius.circular(32.0),
                gapPadding: 0.0,
             ),
            ),
          ),
```

Screen Shot 2021-11-02 at 3 31 14 PM

With this PR, the same code gives the correct result:

Screen Shot 2021-11-02 at 3 30 25 PM

### Todos
  - [x] Test.
  - [x] Make sure the current math doesn't break any other cases.
  - [x] Check other code paths and make sure they don't need their math updated as well.

### Related issues

Fixes https://github.com/flutter/flutter/issues/82321
Old PR that may have caused this bug: https://github.com/flutter/flutter/pull/34515

### CODE REVIEW

File: packages/flutter/lib/src/material/input_border.dart [hunk 1]
```diff
@@ -418,12 +418,11 @@ class OutlineInputBorder extends InputBorder {
 
     const double cornerArcSweep = math.pi / 2.0;
     final double tlCornerArcSweep = start < scaledRRect.tlRadiusX
-      ? math.asin((start / scaledRRect.tlRadiusX).clamp(-1.0, 1.0))
+      ? math.acos(1 - start / scaledRRect.tlRadiusX)
```
Comment (justinmc) [inline packages/flutter/lib/src/material/input_border.dart:421]: Here's the PR where this line originally came from: https://github.com/flutter/flutter/pull/34515. The new calculation appears to work for that case as well.

File: packages/flutter/lib/src/material/input_border.dart [hunk 2]
```diff
@@ -432,7 +431,7 @@ class OutlineInputBorder extends InputBorder {
     const double trCornerArcSweep = cornerArcSweep;
     if (start + extent < scaledRRect.width - scaledRRect.trRadiusX) {
       path
-        ..relativeMoveTo(extent, 0.0)
+        ..moveTo(scaledRRect.left + start + extent, scaledRRect.top)
```
Comment (justinmc) [inline packages/flutter/lib/src/material/input_border.dart:434]: This change and the previous changes near line 425 make the gap the correct width.  Before, it was too wide.

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 1.
View them at https://flutter-gold.skia.org/cl/github/92940

File: packages/flutter/lib/src/material/input_border.dart (hunk 1 above)
Comment (werainkhatri) [reply packages/flutter/lib/src/material/input_border.dart:421]: A suggestion
```dart
final double tlCornerArcSweep = math.asin((1 - start / scaledRRect.tlRadiusX).clamp(0.0, 1.0));
```

**Comment by justinmc** [comment]
The broken goldens are practically unnoticeable 1 pixel differences, so that's good.  I'll plan to just accept those once I finalize this PR.  I should also run the Google tests on this and watch out for breaking other visual diff tests.

File: packages/flutter/lib/src/material/input_border.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/material/input_border.dart:421]: Thanks, I was worried about not having the clamp in there anymore.  Will try it now.

**Comment by werainkhatri** [comment]
I have raised a pull request on your fork with the working solution I wrote. See if that help 😬.

**Comment by skia-gold** [comment]
Gold has detected about 2 new digest(s) on patchset 3.
View them at https://flutter-gold.skia.org/cl/github/92940

**Comment by skia-gold** [comment]
Gold has detected about 4 new digest(s) on patchset 6.
View them at https://flutter-gold.skia.org/cl/github/92940

**Comment by skia-gold** [comment]
Gold has detected about 12 new digest(s) on patchset 8.
View them at https://flutter-gold.skia.org/cl/github/92940

**Comment by skia-gold** [comment]
Gold has detected about 12 new digest(s) on patchset 9.
View them at https://flutter-gold.skia.org/cl/github/92940

**Comment by skia-gold** [comment]
Gold has detected about 1 new digest(s) on patchset 10.
View them at https://flutter-gold.skia.org/cl/github/92940

**Comment by justinmc** [comment]
I've accepted the goldens changes, which as I expected were just two unnoticeable 1 pixel color changes plus the new golden tests I created.

**Comment by justinmc** [comment]
I ran the Google test and got a few failures, but they were all the same as the goldens: a pixel or two that's a slightly different color.  I'll change it to pass and plan to accept the changes after merge.

File: packages/flutter/lib/src/material/input_border.dart [hunk 3]
```diff
@@ -416,28 +416,30 @@ class OutlineInputBorder extends InputBorder {
...
     );
 
+    // This assumes that the radius is circular (x and y radius are equal).
+    // Currently, BorderRadius only supports circular radii.
     const double cornerArcSweep = math.pi / 2.0;
-    final double tlCornerArcSweep = start < scaledRRect.tlRadiusX
-      ? math.asin((start / scaledRRect.tlRadiusX).clamp(-1.0, 1.0))
-      : math.pi / 2.0;
+    final double tlCornerArcSweep = math.acos(
+      (1 - start / scaledRRect.tlRadiusX).clamp(0.0, 1.0),
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/input_border.dart:423]: nit: is it guaranteed that the radiusX can't be 0?

**Review by LongCatIsLooong — APPROVED** [review]

File: packages/flutter/lib/src/material/input_border.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/material/input_border.dart:423]: TIL! So `NaN.clamp(0, 1) == 1`? But shouldn't the `1 - NaN` be evaluated first and throw an error?

Comment (werainkhatri) [reply packages/flutter/lib/src/material/input_border.dart:423]: turns out, `1 - NaN = -NaN`. So it finally clamps to 0.0.

Comment (justinmc) [reply packages/flutter/lib/src/material/input_border.dart:423]: And `1/0` is actually `Infinity` in Dart and not `NaN`, so it makes a little more sense.

**Pull request merged**